_posts/2022-11-23-#26152.md

```diff
@@ -7,17 +7,152 @@ authors: [Xekyo]
+
+    - Notably, each candidate coin is considered using an "effective value," introduced in
+      [PR #17331](https://github.com/bitcoin/bitcoin/pull/17331). This deducts the cost to spend
+      this input at the target feerate from its `nValue`.
+
+- Since [PR #7600](https://github.com/bitcoin/bitcoin/pull/7600), the
+  [`BlockAssembler`](https://github.com/bitcoin/bitcoin/blob/aeb395dcdbfe2b1a6c77ff218939a18afde3add9/src/node/miner.h#L164)
+algorithm has used ancestor feerate, rather than individual feerate, to select transactions for
+inclusion in blocks. This strategy also enables users to fee-bump transactions using [Child Pays for
+Parent (CPFP)](https://bitcoinops.org/en/topics/cpfp/). We have discussed the `BlockAssembler`
+implementation in a [previous review club meeting](/24538).
+
+- On the flip side of CPFP, when creating a transaction at a target feerate, if the wallet selects
+  UTXOs from lower-feerate unconfirmed transactions, it may underestimate the amount of fees to put
+on this transaction. This issue has been an open problem for years; see
+[#9645](https://github.com/bitcoin/bitcoin/issues/9645),
+[#9864](https://github.com/bitcoin/bitcoin/issues/9864), and
+[#15553](https://github.com/bitcoin/bitcoin/issues/15553).
+
+- Goal: when funding a transaction using unconfirmed inputs, in addition to funding the payment(s)
+  and this transaction's fee, also include the fees necessary to bump those unconfirmed transactions
+to the target feerate.
+
+- One naive solution would be to select all the inputs, and then calculate the cost of bumping
+  unconfirmed ancestors and add that to the fees. However, if the selected inputs cannot cover the
+cost of ancestor fees, the selection process needs to be repeated; this strategy would (re)introduce
+a looped coin selection algorithm and make changeless solutions much less common.
+
+- Consider another solution which updates the "effective value" of a coin, reducing it by the fee
+  necessary to bump its ancestor set to the target feerate: `effective_value = nValue -
+cost_to_spend - min(0, target_feerate * ancestor_size - ancestor_fees)`. For example, if the target
+feerate is 20sat/vB, the UTXO has type P2WPKH and value 100ksat, and transaction's ancestor fee and
+vsize are 1200sat and 600vB respectively, the coin's effective value would be 100000sat -
+(20sat/vB * 68vB) - (20sat/vB * 600vB - 1200sat) = 87840sat.
+
+- However, this problem is much more complicated than simply using the ancestor size and fees of
+  unconfirmed UTXOs:
+
+    - Multiple UTXOs may come from the same transaction or share the same ancestry. We don't need to
+      bump those transactions more than once.
+
+    - An ancestor's feerate may be higher than the target feerate.
+
+    - An ancestor may have a low feerate, but be already fee-bumped by another transaction in the
+      mempool.
+
+    - If the transaction may replace something in the mempool, one or multiple transactions may be
```

Comment:
Can this affect ancestors? I'd think this would only evict descendants, and thus not affect ancestor feerate?
```suggestion
    - If the transaction replaces an existing transaction in the mempool, one or multiple transactions may be
```

Reply:
took the grammaf fix and added an example